src/flp/types/dp.rs

```diff
@@ -0,0 +1,331 @@
+use crate::dp::{distributions::PureDpDiscreteLaplace, DifferentialPrivacyStrategy};
+use crate::dp::{DifferentialPrivacyDistribution, DpError};
+use crate::field::{FftFriendlyFieldElement, Field128, Field64};
+use crate::flp::gadgets::{Mul, ParallelSumGadget};
+use crate::flp::types::{Histogram, SumVec};
+use crate::flp::{FlpError, TypeWithNoise};
+use crate::vdaf::xof::SeedStreamTurboShake128;
+use num_bigint::{BigInt, BigUint, TryFromBigIntError};
+use num_integer::Integer;
+use rand::{distributions::Distribution, Rng, SeedableRng};
+
+// TODO(#1071): This is implemented for the concrete fields `Field64` and `Field128` in order to
+// avoid imposing the `BigInt: From<F::Integer>` bound on all callers. In the future, we may want to
+// impose this bound on `FieldElementWithInteger`, which would allow simple blanket implementations
+// here.
+impl<S> TypeWithNoise<PureDpDiscreteLaplace> for SumVec<Field64, S>
+where
+    S: ParallelSumGadget<Field64, Mul<Field64>> + Eq + 'static,
+{
+    fn add_noise_to_result(
+        &self,
+        dp_strategy: &PureDpDiscreteLaplace,
+        agg_result: &mut [Self::Field],
+        _num_measurements: usize,
+    ) -> Result<(), FlpError> {
+        self.add_noise(
+            dp_strategy,
+            agg_result,
+            &mut SeedStreamTurboShake128::from_entropy(),
+        )
+    }
+}
+
+impl<S> TypeWithNoise<PureDpDiscreteLaplace> for SumVec<Field128, S>
+where
+    S: ParallelSumGadget<Field128, Mul<Field128>> + Eq + 'static,
+{
+    fn add_noise_to_result(
+        &self,
+        dp_strategy: &PureDpDiscreteLaplace,
+        agg_result: &mut [Self::Field],
+        _num_measurements: usize,
+    ) -> Result<(), FlpError> {
+        self.add_noise(
+            dp_strategy,
+            agg_result,
+            &mut SeedStreamTurboShake128::from_entropy(),
+        )
+    }
+}
+
+impl<F, S> SumVec<F, S>
+where
+    F: FftFriendlyFieldElement,
+    BigInt: From<F::Integer>,
+    F::Integer: TryFrom<BigInt, Error = TryFromBigIntError<BigInt>>,
+{
+    fn add_noise<R>(
+        &self,
+        dp_strategy: &PureDpDiscreteLaplace,
+        agg_result: &mut [F],
+        rng: &mut R,
+    ) -> Result<(), FlpError>
+    where
+        R: Rng,
+    {
+        // Compute the l1-sensitivity of the aggregation function (assuming the substitution-DP
```

Comment:
What's the difference between "general" and "l1-" sensitivity? The former is used in your writeup.

Reply:
I can't find "general" there, did you mean "global sensitivity"? The two are orthogonal. Global sensitivity, as opposed to smooth sensitivity, gives you one value for your query function, and takes the supremum over all pairs of neighboring datasets. IIRC, smooth sensitivity is parameterized by both the query function and the dataset upon which you evaluate it. The benefit of smooth sensitivity is that you can avoid bad utility due to steep changes if your data is never or rarely in the neighborhood of such issues, but the downside is that you now have to think about the sensitivity of your sensitivity, since the smooth sensitivity is data-dependent. L1-sensitivity and L2-sensitivity just refer to whether you use the taxicab metric or Euclidean metric to measure the distance between query results.

Reply:
Yes I meant global sensitivity.

This context is useful. I first read this as a discrepancy between what the write up says and what the comments say.